src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -5946,13 +5946,24 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
             //
             // When the debugger is synchronizing, trying to perform a GC could deadlock. The GC has the
             // threadstore lock and synchronization cannot complete until the debugger can get the
-            // threadstore lock. However the GC can not complete until it sends the BeforeGarbageCollection
-            // event, and the event can not be sent until the debugger is synchronized. In order to break
-            // this deadlock cycle the GC must give up the threadstore lock, allow the debugger to synchronize,
-            // then try again.
+            // threadstore lock. However, if GC events are enabled, the GC can not complete until it sends the
+            // BeforeGarbageCollection event, and the event can not be sent until the debugger is synchronized. In order to
+            // break this deadlock cycle the GC must give up the threadstore lock, allow the debugger to synchronize,
+            // then try again. At the same time, currently a thread in a forbid-suspend-for-debugger region cannot suspend for
+            // the debugger at this point. To break both deadlocks, sending GC events would be deprecated and typically
+            // AreGarbageCollectionEventsEnabledForNextSuspension() would be false.
             // 
-            || (CORDebuggerAttached() &&
-                (g_pDebugInterface->ThreadsAtUnsafePlaces() || g_pDebugInterface->IsSynchronizing()))
+            ||
+            (
+                CORDebuggerAttached() &&
+                (
+                    g_pDebugInterface->ThreadsAtUnsafePlaces() ||
+                    (
+                        g_pDebugInterface->AreGarbageCollectionEventsEnabledForNextSuspension() &&
```

Comment:
This is adding complexity to preserve data breakpoints in VS versions without our pending fix.  However for these versions the deadlock would still exist.

I wonder if the right fix is to simply never allow `EnableGCNotification(true)`.  We could do that by returning an error when when this is set to true. (Or by removing the interface)

The benefit would be that the deadlock would be fixed completely (at the sake of broken data breakpoints on non-updated VS versions). I guess to me that feels like a better solution.  I would prefer a hard failure on data breakpoints, then a intermittent deadlock.  

@chuckries @gregg-miskelly opinions? ramifications?

Reply:
@sdmaclea @gregg-miskelly I don't have a problem with deprecating EnableGCNotifications entirely, however there is a problem in VS: It appears we never actually check the HRESULT from `AdviseGCEvents(...)`. This means that for existing version of VS, a failed return won't actually fail the data bp code path, and the thread contexts will still get updated to set the hardware data bp. Without the GC events actually enabled, these data bp's would be enabled across GC's and almost certainly cause major issues.

Reply:
The deadlock introduced by https://github.com/dotnet/runtime/pull/40060 in adding the forbid-suspend-for-debugger region is only applicable to .NET 5, for 3.1 a different solution was used and that deadlock is not applicable there. So after this change, that deadlock could still occur in .NET 5 (but not 3.1) when using a VS version that uses GC events.

Not sure I followed the above completely. If the target runtime supports data BPs, is there a suggestion for a change?

Reply:
I can draft the change.